### DISCUSSION TITLE
How to create a unique constraint like in AgensGraph?

### DISCUSSION THREAD

**Discussion by bravius** [apache/age#199] [Q&A]
In AgensGraph I could create a unique property constraint using this syntax:

`CREATE CONSTRAINT ON account ASSERT email IS UNIQUE;`

What is the equivalent in Apache AGE for creating a unique constraint on a property?

**Comment by pdpotter** [comment]
There has been some discussion on this subject in https://github.com/apache/incubator-age/issues/45. There has recently been a commit (https://github.com/apache/incubator-age/commit/57e11a3e6952d0dcfa50de3b92e07fb596df68b0) that fixes an issue with updating the indices.

To create a unique constraint, you first need to create an immutable function (a requirement for creating an index)
```
CREATE OR REPLACE FUNCTION get_email(properties agtype) 
  RETURNS agtype
AS
$BODY$
    select agtype_access_operator($1, '"email"');
$BODY$
LANGUAGE sql
IMMUTABLE;
```
and then use this function for the creation of the unique constraint
```
CREATE UNIQUE INDEX account_email_idx ON mygraph.account(get_email(properties)) ;
```

**Reply by bravius** [reply]
@pdpotter

I solved the problem by creating an immutable function, in order to cast from agtype to jsonb:

`CREATE function dev.agtype_to_jsonb(agtype) 
RETURNS jsonb 
LANGUAGE plpgsql IMMUTABLE
AS 
$$ BEGIN RETURN $1::varchar::jsonb; END; $$`

I could then use the cast function to create indexes using standard json operators:

`CREATE UNIQUE INDEX account_email_index ON account((lower(dev.agtype_to_jsonb(properties)->>'email')))`

So far working great.

Cheers and thanks!

**Reply by pdpotter** [reply]
Nice! Thanks for sharing your improved solution, I will probably use it as well.

**Reply by JoshInnis** [reply]
The root cause of this seems to be that the agtype_access_operator is not IMMUTABLE correct? If that was the case, there would be no need to typecast to JsonB or create helper functions that use the access operator.

**Reply by bravius** [reply]
If I try this I get "_ERROR: operator does not exist: agtype ->> unknown_":

`CREATE UNIQUE INDEX account_email_index ON account((lower(properties->>'email')))`

If I explicitly cast to jsonb instead I get "_ERROR: cannot cast type agtype to jsonb_":

`CREATE UNIQUE INDEX account_email_index ON account((lower(properties::jsonb->>'email')))`

If I cast to varchar then jsonb I get "_ERROR: functions in index expression must be marked IMMUTABLE_":

`CREATE UNIQUE INDEX account_email_index ON account((lower(properties::varchar::jsonb->>'email')))`

The only way I've been able to get it to work is using a custom cast function, such as the one I mentioned above. Then all the usual json operators can be used for indexing.

**Reply by JoshInnis** [reply]
Yes, we are working on a patch to improve some performance issues, part of those fixes was to make agtype_access_operator IMMUTABLE, this is was intended to allow Index scans to be usable on property values in a cypher query. However, it seems that this change will affect this too.

**Reply by JoshInnis** [reply]
Also it appears that creating the  `->` `->>` `#>` `#>>` operators for Agtype is desired as well, and functionality to cast agtype to JsonB

**Comment by emotionbug** [comment]
Looking at this, I hope it have following function.
1. Agtype access operator ( `->` `->>` `#>` `#>>` )
2. Agtype to Jsonb casting function.

**Reply by JoshInnis** [reply]
Isn't all the functionality offered by  `->` `->>` `#>` `#>>` contained with agtype_access_operator? So making those operators would mostly be making helper functions/operators around agtype_access_operator

**Reply by emotionbug** [reply]
Yes, that's right, but the function is not listed in the document. therefore, there is a problem that users do not know how to approach.

As an alternative, the above operators are PostgreSQL's approach to JSON(B), so I think users can use them naturally. Like the case that the above users wrote.

It's already been worked on, so I'll leave it as PR

**Reply by JoshInnis** [reply]
Perhaps this should be added to the documentation if users are not sure how to do it.

**Reply by JoshInnis** [reply]
One issue to be discussed where these operators could be potentially different from the access operator is how they handle vertices, edges, and paths. Should `MATCH (n) WHERE n->'prop_name' = 'value'`  work the same way  `MATCH (n) WHERE n.'prop_name' = 'value'` where the properties of the vertex are used or should the user need to enter  `MATCH (n) WHERE n->'properties'->'prop_name' = 'value'`. Personally I believe the first way is better as it is consistent with the `.` operator.

**Comment by bravius** [comment]
The new agtype access operators `->` and `->>` work great for indexing, thanks for adding those.
